lib/compiler/passes/generate-js.js

```diff
@@ -351,6 +352,50 @@ function generateJS(ast, options) {
         parts.push("}");
       }
 
+      /*
+        MATCH_* opcodes typically do something like
+
+          if (<test>(input.substr(peg$currPos, length))) {
+            sN = input.substr(peg$currPos, length);
+            ...
+          } else {
+            sN = peg$FAILED;
+            ...
+          }
+
+        compileInputChunkCondition will convert that to
+
+          sN = input.substr(peg$currPos, length);
+          if (<test>(sN)) {
+            ...
+          } else {
+            sN = peg$FAILED;
+            ...
+          }
+
+          and avoid extracting the sub string twice.
+      */
+      function compileInputChunkCondition(
+        cond, argCount, inputChunkLength
+      ) {
+        const baseLength = argCount + 3;
+        let inputChunk = inputChunkLength === 1
+          ? "input.charAt(peg$currPos)"
+          : "input.substr(peg$currPos, " + inputChunkLength + ")";
+        if (bc[ip + baseLength] === op.ACCEPT_N
+              && bc[ip + baseLength + 1] === inputChunkLength) {
+          skipAccept = true;
```

Comment:
I think, it should be possible just consume 2 opcodes at once instead of introducing this flag. If you recognize pattern, just pretend that you have a new long opcode and process it

Reply:
I'm going to push back on this one. The only way I see to do it is to put a modified copy of compileCondition into compileInputChunkCondition. Or maybe pass in some extra options to tell it to push a particular string at the beginning the then block, and then skip the first two bytecodes. Either way, I think the current solution is cleaner.

If you have other ideas, or would like me to do one of the above, let me know.

Reply:
> Or maybe pass in some extra options to tell it to push a particular string at the beginning the then block, and then skip the first two bytecodes.

Yes, that is what I mean. Basically you pretend that instead of this opcode:
```js
[
  op.MATCH_STRING, 0,  // string 0
  if_len, else_len,
  op.ACCEPT_N, 2, ..., // if part
  ...                  // else part
]
```
you process this opcode:
```js
[
  /* virtual opcode */,
  if_len-2, else_len,
  ..., // if part
  ...  // else part
]
```
So it seems that all that you need it is correctly calculate IF part to pass to recursion. It seems that you just need add a new parameter to `compileCondition`:
```diff
-function compileCondition(cond, argCount) {
+function compileCondition(cond, argCount, thenSkip) {
          () => {
-           ip += baseLength;
+           ip += baseLength + thenSkip;
            thenCode = compile(bc.slice(ip, ip + thenLength));
            ip += thenLength;
          },
}
```
Pass `0` is all existing invocations and `2` (or whatever) in this new code (which is on line 391 currently)

Reply:
That doesn't work, because It's only supposed to skip a part of the ACCEPT_N.

You also need to tell it to increment peg$currPos by the correct amount at the beginning of the then, and to bump stack.sp, to get things properly balanced...

And passing in a flag to do all that seems like way too much magic. I suppose it could be a callback, but even that feels much worse to me than the current solution.

Reply:
I'm actually working on some bytecode optimizations that will produce lots more MATCH combinations by pulling in subsequent bytecode.

That is to say that where currently we get
```
MATCH
  ACCEPT
ELSE
  FAIL
IF_FAILED // or IF_NOT_FAILED
  DO_FAILED_THINGS
ELSE
  DO_MATCH_THINGS
```
My optimization converts that to

```
MATCH
  ACCEPT
  DO_MATCH_THINGS
ELSE
  FAIL
  DO_FAILED_THINGS
```
This both saves the second conditional, and in many cases allows other cleanups - often it can avoid assigning the result of the match to anything, or assigning peg$FAILED to anything.

So I think this approach (which continues to work, and can be expanded after my next pull request), is the way to go, rather than pulling out MATCH_ACCEPT patterns.

Reply:
Thinking about it, if you really hate the `skipAccept` flag, I think I can make a callback work fairly cleanly:

```
     ip += baseLength;
     thenCode = (callback || compile)(bc.slice(ip, ip + thenLength));
     ip += thenLength;
```

Then existing callers of compileCondition can be left unchanged, and compileInputChunkCondition can pass in a suitable callback to emulate the bits of accept that it needs to do, and skip over the first couple of bytecodes.

Let me know if you'd like me to do something like that.

Reply:
No, this is not necessary, but if you like to try this approach I also wouldn't be against

Reply:
I tried, and I mostly like it, so I've pushed the update. The only downside is that now the code to handle ACCEPT_N is split into two places - the actual ACCEPT_N, and the cut down version handled by the callback to compileCondition. 

But it avoids the statefulness of skipAccept, so I think its a good change on balance.